docs/designs/custom_allreduce_training_loop_support.md

```diff
@@ -0,0 +1,402 @@
+
+![Training Loop](../images/training_loop.jpg)
+
+In order to support customized training loop with ElasticDL AllReduce,
+ElasticDL provides users with a dataset with dynamic partition support,
+and an API `elastic_run` for elastic training. Users need to implement
+their own training logic for a training data batch in a function
+`train_one_batch`.
+
+```Python
+def train_one_batch(batch_data):
+    loss = forward(batch_data)
+    gradients = backward(loss)
+    gradients = allreduce(gradients)
+    upate_model(gradients)
+```
+
+Then users can write the training loop using `train_one_batch`,
+the dataset and `elastic_run` API provided by ElasticDL.
+
+```python
+def train(dataset, elastic_manager):
+    """
+    Arguments:
+        dataset: tf.data.Dataset. ElasticDL creates a dataset according to
+            tasks from the master
+        elastic_manager: A manager to execute elastic training.
+    """
+    for batch_data in dataset:
+        elastic_train = elastic_manager.elastic_run(train_one_batch)
+        elastic_train(batch_data)
+```
+
+## Elastic AllReduce Wrapper for Customized Training Loop
+
+ElasticDL AllReduce is based on Horovod. Using Horovod, users only need to
+use Horovod APIs to wrap optimizer or tape to execute AllReduce in their
+training loop.
+
+```python
+tape = hvd.DistributedGradientTape(tape)
+grads = tape.gradient(loss, model.trainable_variables)
+optimizer.apply_gradients(
+    zip(grads, model.trainable_variables)
+)
+```
+
+We can develop a decorator to wrap the gradient combination process
+using AllReduce. The decorator should support fault-tolerance and
+elastic training. If AllReduce fails, the decorator will query the
+master for a new AllReduce ring and retry to combine gradients
+across alive workers. And the decorator also queries the master
+for a new ring periodically if the master adds new workers.
+
+```python
+class ElasticAllReduceManager(object):
+    def elastic_run(self, func):
+        def wrapper(*args, **kwargs):
+            # In case where the master add new workers
+            if ITER_STEP % STEPS_TO_CHECK_RING == 0:
+                init_horovod_if_needed()
+
+            for i in range(MAX_ALLREDUCE_RETRY_COUNT):
+                # Retry to allreduce if failed.
+                try:
+                    func(*args, **kwargs)
+                    report_batch_finished()
+                except HorovodInternalError:
+                    init_horovod_if_needed()
+
+                ITER_STEP += 1
+        return wrapper
+
+    def broadcast(self):
+        broadcast_model()
+        broadcast_optimizer()
+
+    def set_broadcast_object(self, broadcast_obj):
+        ...
+```
+
+## Examples of Different Libraries Using ElasticDL AllReduce
+
+Users only need to make the following changes to their training loop:
+
+- Use the decorator of elastic training to wrap the code of gradient
+combination using Horovod.
+- Set objects to broadcast like the model and optimizer. Because an alive
```

Comment:
How to `set_broadcast_object`? Any example?

Reply:
I have listed the example in the following code snippets like:

```python
# Set object to broadcast
elastic_manager.set_model(mnist_model)
elastic_manager.set_optimizer(opt)
```